source/Pages/jupyter_notebooks.rst

```diff
@@ -7,6 +8,147 @@ Jupyter Notebooks
+
+You can create a virtual environment (or venv) at a path by doing:
+
+.. code-block:: bash
+   
+   python3.9 -m venv test_venv/
+
+This will create a folder called ``test_venv`` which contains the entire python environment. You can also use other python versions if you prefer. To load this environment run:
+
+.. code-block:: bash
+   
+   source test_venv/bin/activate
+
+This will show in some shells a ``(test_venv)`` next to your command line. In the environment you can now install packages using pip:
+
+.. code-block:: bash
+   
+   pip install jupyterlab pandas uproot4
+
+To start a jupyter session, run
+
+.. code-block:: bash
+   
+   jupyter lab --ip="*" --no-browser
+
+Where the ``ip`` flag and the ``no-browser`` respectively ensure that the session is forwarded through the network and that no browser is opened in an ``X11`` session that may be running through your ``ssh`` connection.
+
+To properly forward the lab session to your local machine, a second terminal has to be opened running:
+
+.. code-block:: bash
+   
+   ssh -NL 8888:wn-db-06:8888 spider
+
+where the machine name has to match where the kernel is running (``wn-db-06`` has to match) and the forwarded port (in this example ``8888``) has to match the port given by the jupyter-lab instance. Again, **do not run notebooks on UI machines**. Now that the tunnel is opened and should forward the connection to your browser, open the link provided by jupyter in your favorite browser. The link has the shape ``http://localhost:8888/lab?token=abc123``.
+
+Once you are done with the virtual environment and want to go back to the inital user environment type:
+
+.. code-block:: bash
+
+   deactivate
+
+and the python environment is unloaded.
+
+.. WARNING::
+   Some jupyter instances provide a link of that contains ``hostname:8888``. Replace ``hostname`` with ``localhost`` or ``127.0.0.1`` to properly fetch the notebook.
+
+=====================
+Singularity container
+=====================
+
+Pre-built container
+===================
+
+To run a notebook in a singularity container, we have to fetch or build the container first. A tutorial on containers can be found in :ref:`singularity-building`, but note that this focuses on using GPUs. A more general introduction is provided here.
+
+First we start by fetching a container:
+
+.. code-block:: bash
+
+   singularity build jupyter.sif docker://jupyter/scipy-notebook:latest
+
+This will pull one of the official jupyter containers from docker hub, and build a singularity container from it. This container encapsulates the entire environment and can be entered to start a notebook session. More official containers can be found `here <https://hub.docker.com/r/jupyter/>`_, and more docker images in general can be found at `docker hub <https://hub.docker.com/>`_.
+
+After the build procedure is complete, you can start the jupyter instance with 
```

Comment:
Consider adding a note here on starting jupyter in an interactive node session and not the ui where they may have fetched the image (as in previous section).